src/dev/flang/ir/IR.java

```diff
@@ -337,6 +340,204 @@ else if (e instanceof Universe)
+        var rt = needsBoxing(expr, frmlT);
+        if (rt != null)
+          {
+            result = new Box(result, rt);
+          }
+      }
+
+    if (POSTCONDITIONS) ensure
+      (Errors.any()
+        || t.isVoid()
+        || frmlT.isGenericArgument()
+        || frmlT.isThisType()
+        || needsBoxing(result, frmlT) == null
+        || frmlT.isAssignableFrom(t).no());
+
+    return result;
+  }
+
+
+  /**
+   * handle tagging when assigning value to choice frmlT
+   *
+   * @param expr
+   *
+   * @param frmlT
+   *
+   * @return
+   */
+  private static Expr tag(Expr expr, AbstractType frmlT)
+  {
+    if(PRECONDITIONS) require
+      (frmlT.isChoice());
+
+    // Case 1: types are equal, no tagging necessary
+    if (frmlT.compareTo(expr.type()) == 0)
+      {
+        return expr;
+      }
+    // Case 1.1: types are equal, no tagging necessary
+    // NYI: BUG: soundness issue? see also isAssignableFrom
+    else if(expr.type().isChoice() && (frmlT.isThisType() || expr.type().isThisType()) && frmlT.asThis().compareTo(expr.type().asThis()) == 0)
+      {
+        return expr;
+      }
+    // Case 2.1: ambiguous assignment via subtype
+    //
+    // example:
+    //
+    //  A ref is
+    //  B ref is
+    //  C ref : B, A is
+    //  t choice A B := C
+    //
+    else if (frmlT
+             .choiceGenerics()
+             .stream()
+             .filter(cg -> cg.isAssignableFromWithoutTagging(expr.type()).yes())
+             .count() > 1)
+      {
+        Errors.fatal("Ambiguous assignment to choice, should have been caught in frontend.");
+        return expr;
+      }
+    // Case 2.2: no nested tagging necessary:
+    // there is a choice generic in this choice
+    // that this value is "directly" assignable to
+    else if (frmlT
+             .choiceGenerics()
+             .stream()
+             .anyMatch(cg -> cg.isAssignableFromWithoutTagging(expr.type()).yes()))
+      {
+        return new Tag(expr, frmlT);
+      }
+    // Case 3: nested tagging necessary
+    // value is only assignable to choice element
+    // that itself is a choice
+    else
+      {
+        // we assign to the choice generic
+        // that expr is assignable to
+        var cgs = frmlT
+          .choiceGenerics()
+          .stream()
+          .filter(cg -> cg.isChoice() && cg.isAssignableFromWithoutBoxing(expr.type()).yes())
+          .collect(Collectors.toList());
+
+        if (cgs.size() > 1)
+          {
+            Errors.fatal("Ambiguous assignment to choice, should have been caught in frontend.");
+          }
+
+        if (CHECKS) check
+          (Errors.any() || cgs.size() == 1);
+
+        return tag(tag(expr, cgs.get(0)), frmlT);
+      }
+  }
+
+
+  /**
+   * Is boxing needed when we assign to frmlT?
+   *
+   * @param frmlT the formal type we are assigning to.
+   *
+   * @return the type after boxing or null if boxing is not needed
+   */
+  private static AbstractType needsBoxing(Expr expr, AbstractType frmlT)
```

Comment:
This appears to do the same as `Expr.boxedType()`, but slightly differently. It would IMHO be good to move this to `Expr` and to merge `boxedType` and `needsBoxing` into one method or to move the core logic to a common helper method. 

Or, since `Expr.boxedType()` seems to be used only in `Expr.checkAmbiguousAssignmentToChoice`, maybe we can avoid `Expr.boxedType()` altogether and use `needsBoxing` instead for this check?